colossus/src/main/scala/colossus/protocols/http/HttpResponse.scala

```diff
@@ -1,77 +1,151 @@
 package colossus
 package protocols.http
 
-import akka.util.ByteString
-import colossus.controller.{FiniteBytePipe, Source}
+import akka.util.{ByteString, ByteStringBuilder}
+import colossus.controller.Source
 import colossus.core.DataBuffer
+import core._
+import service.Callback
+import controller._
+import HttpParse._
+
+
+case class HttpResponseHead(version : HttpVersion, code : HttpCode, headers : Vector[(String, String)] = Vector()) extends HttpHeaderUtils {
+
+  def appendHeaderBytes(builder : ByteStringBuilder) {
+    builder append version.messageBytes
+    builder putByte ' '
+    builder append code.headerBytes
+    builder append NEWLINE
+    headers.foreach{case (key, value) =>
+      builder putBytes key.getBytes
+      builder putBytes ": ".getBytes
+      builder putBytes value.getBytes
+      builder append NEWLINE
+    }
+  }
 
-trait HttpResponseHeader extends HttpHeaderUtils{
-  def version : HttpVersion
-  def code : HttpCode
-  def headers : Seq[(String, String)]
+  def withHeader(key: String, value: String) = copy(headers = headers :+ (key, value))
 
 }
 
-case class HttpResponse(version : HttpVersion, code : HttpCode, headers : Seq[(String, String)] = Nil, data : ByteString) extends HttpResponseHeader {
+sealed trait BaseHttpResponse { 
 
-  def withHeader(key: String, value: String) = copy(headers = headers :+ (key, value))
+  type Encoded <: DataReader
 
-  override def equals(other: Any) = other match {
-    case HttpResponse(v, c, h, b) => v == version && c == code && b == data && h.toSet == headers.toSet
-    case _ => false
+  def head: HttpResponseHead
+
+  /**
+   * Resolves the body into a bytestring.  This operation completes immediately
+   * for static responses.  For streaming responses, this only completes when
+   * all data has been received.  Be aware if the response is an infinite
+   * stream of data (using chunked transfer encoding), this will never
+   * complete.
+   */
+  def resolveBody(): Option[Callback[ByteString]] = this match {
```

Comment:
Indifferent remark: Any reason why this is just not abstract with each subclass providing their own impl?